### ISSUE TITLE
number_format() expects parameter 1 to be float, string given

### ISSUE THREAD

**Issue by adi2k5** [pi-hole/AdminLTE#647]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/AdminLTE/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/AdminLTE/issues)).

**How familiar are you with the the source code relevant to this issue?:**
3
---
**Expected behaviour:**

`See all data on the webpage`

**Actual behaviour:**

`'small boxes' are not showing any value but just '---', in php log I can see` 
```
[17-Dec-2017 13:44:13 Europe/Warsaw] PHP Warning:  number_format() expects parameter 1 to be float, string given in /var/www/html/admin/api_FTL.php on line 39
```
`warning. Beside this, all is working fine, all charts, statistics of top blocked domains etc. are being shown`

**Steps to reproduce:**

`install and run pi-hole webpage`

**Troubleshooting undertaken, and/or other relevant information:**

`This is a follow up of discussion from issue` https://github.com/pi-hole/pi-hole/issues/1832

lighttpd log:
```
192.168.1.1 my.domain user [17/Dec/2017:11:58:50 +0100] "GET /admin/api.php?summary HTTP/1.1" 200 174 "https://my.domain/admin/index.php?login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.97 Safari/537.36 Vivaldi/1.94.1008.34"
192.168.1.1 my.domain user [17/Dec/2017:11:58:51 +0100] "GET /admin/api.php?summaryRaw&topItems HTTP/1.1" 200 325 "https://my.domain/admin/index.php?login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.97 Safari/537.36 Vivaldi/1.94.1008.34"
192.168.1.1 my.domain user [17/Dec/2017:11:58:51 +0100] "GET /admin/api.php?summaryRaw&getQuerySources HTTP/1.1" 200 197 "https://my.domain/admin/index.php?login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.97 Safari/537.36 Vivaldi/1.94.1008.34"
192.168.1.1 my.domain user [17/Dec/2017:11:58:51 +0100] "GET /admin/api.php?summary HTTP/1.1" 200 175 "https://my.domain/admin/index.php?login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.97 Safari/537.36 Vivaldi/1.94.1008.34"
```
As suggested in https://github.com/pi-hole/pi-hole/issues/1832 it might be locale related issue, in Poland we're using comma as decimal separator and thin space as thousands groups separator.

**Comment by DL6ER** [comment]
Okay, as you see [here](https://github.com/pi-hole/pi-hole/issues/1832#issuecomment-352241257), `FTL` *always* uses a `.` as decimal separator for reliability
```
ads_percentage_today 42.622002
```
Apparently, `PHP` uses the computer's locale and expects `.` as thousand separator and `,` as decimal separator.

> Beside this, all is working fine, all charts, statistics of top blocked domains etc. are being shown.

Okay, this is good to know. I think we can just remove `number_format()` altogether and send pre-formated strings from `FTL` to the API. Asking @pi-hole/api for their opinion.

**Comment by Mcat12** [comment]
It would be preferable to have the API output consistent regardless of the server's locale. Also, changing it to output strings instead of numbers would affect anything else using it (web interface, chronometer, custom user scripts, etc).
I'm still a bit confused as to why PHP is having such a hard time, because the docs don't mention locale affecting `number_format`:
```php
$number = 1234.56;

// english notation (default)
$english_format_number = number_format($number);
// 1,235
```

The default parameters seem to indicate that it keeps the English notation as the default even through locale changes:
```php
string number_format(float $number, int $decimals = 0, string $dec_point = ".", string $thousands_sep = ",")
```

@adi2k5 your warning line number still seems off to me. What is on line 39 of your `/var/www/html/admin/api_FTL.php`?

**Comment by DL6ER** [comment]
Hmm...
```
<?php
$number = 1234.56;

function print_loc($loc)
{
	global $number;
	$loc = setlocale(LC_ALL, $loc);
	echo $loc.": ".number_format($number,2)."\n";
	echo strftime("%A %d %B %Y", mktime(0, 0, 0, 12, 22, 1978))."\n";
}

print_loc("en_US.UTF-8");
print_loc("de_DE.UTF-8");
print_loc("pl_PL.UTF-8");
?>
```
results in
```
en_US.UTF-8: 1,234.56
Friday 22 December 1978

de_DE.UTF-8: 1,234.56
Freitag 22 Dezember 1978

pl_PL.UTF-8: 1,234.56
piątek 22 grudzień 1978
```
so it doesn't actually seem that `number_format()` is depending on the locale

**Comment by adi2k5** [comment]
@Mcat12 it's the same as in repo:
```
# pwd; sed -n '39,39p' api_FTL.php
/var/www/html/admin
                        continue;
# 
```
I just run dev tools, and from what I can see, each second (!) following request is being called
`jquery.min.js:4 XHR finished loading: GET "https://my.domain/admin/api.php?summary".`
and this request generate warning above. 
Checking what kind of data is under above link, I have:
`{"domains_being_blocked":"2,852,588","dns_queries_today":"22,663","ads_blocked_today":"17,216","ads_percentage_today":"76.0","unique_domains":"964","queries_forwarded":"4,122","queries_cached":"1,173","clients_ever_seen":"3","unique_clients":"3","status":null}`
I'm not sure if it might be anyhow related, but status here is null, while call to api.php without `?summary` return status:0, so my guess woud be to look more closely at this, as I believe this could be our 'string' instead of 'float' value.

**Comment by DL6ER** [comment]
> I'm not sure if it might be anyhow related, but status here is null, while call to api.php without ?summary return status:0, so my guess woud be to look more closely at this, as I believe this could be our 'string' instead of 'float' value.

`number_format()` should not be fired for `status` as there is an exception for this here:
https://github.com/pi-hole/AdminLTE/blob/master/api_FTL.php#L36

**Comment by Mcat12** [comment]
The short-circuit rules might cause the if statement to fail if only the first statement is false:
```php
if(($tmp[0] === "domains_being_blocked" && !is_numeric($tmp[1])) || $tmp[0] === "status")
```

However, this seems a bit implausible... More things should be going wrong then.

**Comment by Mcat12** [comment]
> call to api.php without ?summary return status:0

It should return a string: `"status":"enabled"`. Does it return a number instead?

**Comment by adi2k5** [comment]
I'm getting different responses for different queries. 
For api.php
```
{"domains_being_blocked":2853322,"dns_queries_today":6436,"ads_blocked_today":2686,"ads_percentage_today":41.733997,"unique_domains":1157,"queries_forwarded":2949,"queries_cached":736,"clients_ever_seen":3,"unique_clients":1,"status":0}
```
for api.php?summary
```
{"domains_being_blocked":"2,853,322","dns_queries_today":"6,459","ads_blocked_today":"2,694","ads_percentage_today":"41.7","unique_domains":"1,157","queries_forwarded":"2,961","queries_cached":"739","clients_ever_seen":"3","unique_clients":"1","status":null}
```
for api.php?summaryRaw
```
{"domains_being_blocked":2853322,"dns_queries_today":6465,"ads_blocked_today":2696,"ads_percentage_today":41.701469,"unique_domains":1157,"queries_forwarded":2963,"queries_cached":741,"clients_ever_seen":3,"unique_clients":1,"status":0}
```
and for api.php?status
```
{"status":"enabled"}
```

**Comment by Mcat12** [comment]
What is the output of:
```
echo ">stats" | nc localhost 4711
```

**Comment by adi2k5** [comment]
Sorry for delay, here's what I answered already in https://github.com/pi-hole/pi-hole/issues/1832 , except numbers, nothing else changed.
```
echo ">stats" | nc localhost 4711
domains_being_blocked 2852587
dns_queries_today 8627
ads_blocked_today 3677
ads_percentage_today 42.622002
unique_domains 983
queries_forwarded 4050
queries_cached 876
clients_ever_seen 1
unique_clients 1
status enabled
---EOM---
```

**Comment by Mcat12** [comment]
And you still see this issue?

**Comment by adi2k5** [comment]
Yep, just updated to 
```
# pihole version
  Pi-hole version is v3.2.1 (Latest: v3.2.1)
  AdminLTE version is v3.2.1 (Latest: v3.2.1)
  FTL version is vDev-437af07 (Latest: v2.13.1)
```
and it's still there.

**Comment by Mcat12** [comment]
Try another `pihole -up`, because there's an FTL update with some bugfixes, including fixing the version display.

I will manually test with the data you shared, but I think at this point we will have to do some manual `print` debugging. (I'll tell you where to put the debugging lines soon)

**Comment by adi2k5** [comment]
```
# pihole version
  Pi-hole version is v3.2.1 (Latest: v3.2.1)
  AdminLTE version is v3.2.1 (Latest: v3.2.1)
  FTL version is v2.13.1 (Latest: v2.13.1)
```
Still no values but '---' in blocks. 
One thing I forgot to mention earlier is, that it was working normally while I was on 3.1, problem occured only after update to 3.2.

**Comment by Mcat12** [comment]
On [line 30 of api_FTL.php](https://github.com/pi-hole/AdminLTE/blob/master/api_FTL.php#L30) add this command:
```php
var_dump($return);
```

and share the output of those 3 API calls you shared earlier.

**Comment by adi2k5** [comment]
Hmmm, it doesn't seem to be changing anything. Api.php:
```
{"domains_being_blocked":2860442,"dns_queries_today":18388,"ads_blocked_today":10286,"ads_percentage_today":55.938656,"unique_domains":1059,"queries_forwarded":7261,"queries_cached":841,"clients_ever_seen":1,"unique_clients":1,"status":0}
```
api.php?summary
```
{"domains_being_blocked":"2,860,442","dns_queries_today":"18,422","ads_blocked_today":"10,295","ads_percentage_today":"55.9","unique_domains":"1,059","queries_forwarded":"7,284","queries_cached":"843","clients_ever_seen":"1","unique_clients":"1","status":null}
```
api.php?summaryRaw
```
{"domains_being_blocked":2860442,"dns_queries_today":18423,"ads_blocked_today":10295,"ads_percentage_today":55.881237,"unique_domains":1059,"queries_forwarded":7285,"queries_cached":843,"clients_ever_seen":1,"unique_clients":1,"status":0}
```
api.php?status
```
{"status":"enabled"}
```
```
# grep -nC1 var_dump api_FTL.php 
29-     $return = getResponseFTL();
30:     var_dump($return);
31-     $stats = [];
```
I've enabled also error reporting in php, but unfortunately it won't give any useful info in api.php?summary beside already known warning
```
<br />
<b>Warning</b>:  number_format() expects parameter 1 to be float, string given in <b>/var/www/html/admin/api_FTL.php</b> on line <b>39</b><br />
{"domains_being_blocked":"2,860,442","dns_queries_today":"18,524","ads_blocked_today":"10,326","ads_percentage_today":"55.7","unique_domains":"1,061","queries_forwarded":"7,348","queries_cached":"850","clients_ever_seen":"1","unique_clients":"1","status":null}
```

**Comment by DL6ER** [comment]
That is *very* strange. With this change,
```
$ grep -nC1 var_dump api_FTL.php 
29-     $return = getResponseFTL();
30:     var_dump($return);
31-     $stats = [];
```
I get
```
array(10) {
  [0]=>
  string(28) "domains_being_blocked 100835"
  [1]=>
  string(22) "dns_queries_today 7615"
  [2]=>
  string(22) "ads_blocked_today 1068"
  [3]=>
  string(30) "ads_percentage_today 14.024951"
  [4]=>
  string(18) "unique_domains 684"
  [5]=>
  string(22) "queries_forwarded 3794"
  [6]=>
  string(19) "queries_cached 2753"
  [7]=>
  string(19) "clients_ever_seen 8"
  [8]=>
  string(16) "unique_clients 7"
  [9]=>
  string(14) "status enabled"
}
{"domains_being_blocked":100835,"dns_queries_today":7615,"ads_blocked_today":1068,"ads_percentage_today":14.024951,"unique_domains":684,"queries_forwarded":3794,"queries_cached":2753,"clients_ever_seen":8,"unique_clients":7,"status":"enabled"}
```
which is what is expected after adding `var_dump();`

**Comment by DL6ER** [comment]
Anyhow, I think it makes more sense to put it in a bit later, like this:
```
if (isset($_GET['summary']) || isset($_GET['summaryRaw']) || !count($_GET))
{
	sendRequestFTL("stats");
	$return = getResponseFTL();

	$stats = [];
	foreach($return as $line)
	{
		$tmp = explode(" ",$line);
		var_dump($tmp); // <----------------- here
[...]
```

**Comment by Mcat12** [comment]
@adi2k5 Do you have another SD card you can try re-installing onto? It might be an issue with your SD card.

**Comment by adi2k5** [comment]
Sooo...sorry for wasting your time. I have finally figured out what's wrong, and fortunately it's not pi-hole problem, so this issue can be closed. After I couldn't get any response from _any_ var_dump, I've created new file with some example var_dump. This one worked, but when I've changed values,  var_dump was still showing old ones, which led me to dig more into server cache configuration (refreshing/removing client cache as suggested in other issue weren't helping), and... bingo. After removing pihole content from /var/cache/lighttpd pihole finally show whole page as expected and warning dissapeared. Thank you once more for your time on this, donated few $.
